### ISSUE TITLE
Path is not PathLike

### ISSUE THREAD

**Issue by vassilyl** [microsoft/pyright#640]
Cannot use `pathlib.Path` as a value for `os.PathLike`. 

**To Reproduce**
```python
from os import PathLike
from pathlib import Path

def check(p: PathLike):
    print(isinstance(p, PathLike))

check(Path())
```
This is a valid script that prints `True`. Although `pyright` v1.1.34 considers this an error:
```
  7:7 - error: Argument of type "Path" cannot be assigned to parameter "p" of type "_PathLike[Union[str, bytes]]"
  TypeVar "AnyStr" is invariant
    Type "str" cannot be assigned to type "Union[str, bytes]" (reportGeneralTypeIssues)
1 error, 0 warnings
```

**Expected behavior**
No error.

**Comment by erictraut** [comment]
I think Pyright is doing the right thing here. Path is not a subclass of PathLike. 

The only reason isinstance returns true in this case is because of a hack in the PathLike implementation where it overrides the `__subclasshook__` to act as though `Path` is a subclass of `PathLike` even though it's not.

The correct fix for this is for the typeshed stub to define PathLike as a Protocol rather than a normal class. There's already a bug logged on this issue in typeshed: https://github.com/python/typeshed/issues/2808

As a workaround in the meantime, you could change your code as follows:

```
check(bytes(Path())
```
This is possible because the `Path` class contains a method `__bytes__`, which returns a `bytes` object, and this is compatible with `PathLike`.

**Comment by vassilyl** [comment]
> Path is not a subclass of PathLike

No, it is not in `pathlib.py` in the standard library itself. But it _is_ in [typeshed](https://github.com/python/typeshed/blob/f7d240f06e5608a20b2daac4e96fe085c0577239/stdlib/3/pathlib.pyi#L66), where it becomes `PathLike[str]` -> `PurePath` -> `Path`.

Let's return to the top post. Correct me if I'm wrong but the Pyright error message means the following.
- The function parameter is of type `_PathLike[AnyStr]` where `AnyStr` is restricted by `Union[str, bytes]`
- The argument value is of type `_PathLike[str]`
- Cannot assign the latter to the former because the first type argument `AnyStr` is invariant and the two types (`Union[str, bytes]` and `str`) must match exactly which is not the case. 

Here Pyright treats generic type `PathLike` without type arguments as if it is an instance type with `AnyStr` bound to `Union[str, byte]`. But the type argument is not bound yet, it is just restricted. In contrast, in `Path` the `AnyStr` type variable _is_ bound to `str`.

As a confirmation of the above, rewriting sample code to `def check(p: PathLike[Any]):`
makes Pyright happy. This is not a valid construct though, it raises TypeError in python. The *real* `PathLike` is not subscriptable.

PEP 484 states:

> Subclassing a generic class without specifying type parameters assumes Any for each position. 

Evidently, Pyright doesn't follow this guidance here. Shall you revisit how Pyright treats generic types without specifying type arguments?

**Comment by erictraut** [comment]
I missed that inheritance hierarchy. Your analysis is spot on, and you're correct that this is a bug. 

PEP 484 contains lots of ambiguities. The line you quoted doesn't make it clear whether this rule applies even in cases where the type var is bound or constrained. I think you're correct that it should still apply in the case of constraints. I've adjusted the code accordingly.

This change will be in the next published version of Pyright.

Thanks for your bug report!

**Comment by erictraut** [comment]
This is now addressed in version 1.1.35, which I just published.
